bn-apps/cordapp-updates-distribution/corda-updates-transport/src/main/kotlin/net/corda/cordaupdates/transport/Transporters.kt

```diff
@@ -0,0 +1,92 @@
+
+import com.sun.xml.internal.messaging.saaj.util.ByteInputStream
+import net.corda.cordaupdates.transport.flows.GetResourceFlow
+import net.corda.cordaupdates.transport.flows.PeekResourceFlow
+import net.corda.core.flows.FlowException
+import net.corda.core.node.AppServiceHub
+import net.corda.core.serialization.CordaSerializable
+import net.corda.core.utilities.getOrThrow
+import org.eclipse.aether.RepositorySystemSession
+import org.eclipse.aether.repository.RemoteRepository
+import org.eclipse.aether.spi.connector.transport.AbstractTransporter
+import org.eclipse.aether.spi.connector.transport.GetTask
+import org.eclipse.aether.spi.connector.transport.PeekTask
+import org.eclipse.aether.spi.connector.transport.PutTask
+import org.eclipse.aether.spi.connector.transport.Transporter
+import org.eclipse.aether.transfer.NoTransporterException
+import org.eclipse.aether.util.ConfigUtils
+
+/**
+ * Transport over Corda flows. Should be used only within a Corda node. This transporter expects an instance of [AppServiceHub]
+ * to be passed via custom session properties. An instance of [AppServiceHub] is required to delegate the data transfer to a separate flow,
+ * to avoid checkpointing of Maven Resolver contents as the they are not @Suspendable.
+ *
+ * When this transport is used from Corda OS, [CordaMavenResolver] should be invoked from a non-flow thread, to prevent the invoking flow from
+ * blocking the only thread available to the flows executor. For example.
+ *
+ * Under the hood [FlowsTransporter] uses [GetResourceFlow] and [PeekResourceFlow] to transfer the content.
+ *
+ * @CordaService
+ * class ResolverExecutor(val appServiceHub : AppServiceHub) : SingletonSerializeAsToken() {
+ *     companion object {
+ *         val EXECUTOR = Executors.newSingleThreadExecutor()
+ *     }
+ *
+ *     fun invoke(resolver : CordaMavenResolver) {
+ *         EXECUTOR.submit {
+ *             val additionalProperties = mapOf(Pair(SessionConfigurationProperties.APP_SERVICE_HUB, appServiceHub))
+ *             resolver.downloadVersion("net.corda:corda-finance", additionalProperties)
+ *         }
+ *     }
+ * }
+ *
+ * class MyFlow(val resolver : CordaMavenResolver) : FlowLogic<Unit>() {
+ *     @Suspendable
+ *     override fun call() {
+ *         val executor = serviceHub.cordaService(ResolverExecutor::class.java)
+ *         executor.invoke(resolver)
+ *     }
+ * }
+ */
+class FlowsTransporter(private val session : RepositorySystemSession,
+                       private val repository : RemoteRepository) : AbstractTransporter() {
+
+    private val repoHosterName = repository.url.substring(repository.protocol!!.length + 1, repository.url.length)
+
+    init {
+        session.configProperties
+        if (repository.protocol.toLowerCase() !in setOf(CordaTransporterFactory.CORDA_FLOWS_TRANSPORT)) {
+            throw NoTransporterException(repository)
+        }
+    }
+
+    override fun implPeek(task : PeekTask?) {
```

Comment:
Kotlin tip: You can just remove the `?` from the type. The compiler will still accept it just fine, and it will insert the null check at the start of the method body for you (invisibly). This means you won't need to use `!!`

Reply:
I didn't know that. Thanks :)